pretyping/pretyping.ml

```diff
@@ -687,30 +716,46 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
-            let value, typ = app_f n (j_val resj) ujval, subst1 ujval c2 in
-	    let j = { uj_val = value; uj_type = typ } in
-            apply_rec env sigma (n+1) j candargs rest
-	  | _ ->
-            let sigma, hj = pretype empty_tycon env sigma c in
-	      error_cant_apply_not_functional
-                ?loc:(Loc.merge_opt floc argloc) !!env sigma resj [|hj|]
+          let sigma, ujval = adjust_evar_source sigma na.binder_name ujval in
+          let value, typ = app_f n (j_val resj) ujval, subst1 ujval c2 in
+          let j = { uj_val = value; uj_type = typ } in
+          apply_rec env sigma (n+1) j candargs bidiargs rest
+        | _ ->
+          let sigma, hj = pretype empty_tycon env sigma c in
+          error_cant_apply_not_functional
+            ?loc:(Loc.merge_opt floc argloc) !!env sigma resj [|hj|]
     in
-    let sigma, resj = apply_rec env sigma 1 fj candargs args in
+    let sigma, resj, bidiargs = apply_rec env sigma 0 fj candargs [] args in
     let sigma, resj =
       match EConstr.kind sigma resj.uj_val with
       | App (f,args) ->
-          if Termops.is_template_polymorphic_ind !!env sigma f then
-	    (* Special case for inductive type applications that must be 
-	       refreshed right away. *)
-            let c = mkApp (f, args) in
-            let sigma, c = Evarsolve.refresh_universes (Some true) !!env sigma c in
-            let t = Retyping.get_type_of !!env sigma c in
-            sigma, make_judge c (* use this for keeping evars: resj.uj_val *) t
-          else sigma, resj
+        if Termops.is_template_polymorphic_ind !!env sigma f then
+          (* Special case for inductive type applications that must be
+             refreshed right away. *)
+          let c = mkApp (f, args) in
+          let sigma, c = Evarsolve.refresh_universes (Some true) !!env sigma c in
+          let t = Retyping.get_type_of !!env sigma c in
+          sigma, make_judge c (* use this for keeping evars: resj.uj_val *) t
+        else sigma, resj
       | _ -> sigma, resj
     in
-    inh_conv_coerce_to_tycon ?loc env sigma resj tycon
+    let app = inh_conv_coerce_to_tycon ?loc env sigma resj tycon in
+    let refine_arg sigma (newarg,origarg) =
+      (* Refine an argument (originally `origarg`) represented by an evar
+         (`newarg`) to use typing information from the context *)
+      (* Recover the expected type of the argument *)
+      let ty = Retyping.get_type_of !!env sigma newarg in
+      (* Type the argument using this expected type *)
+      let sigma, j = pretype (Some ty) env sigma origarg in
+      (* Unify the (possibly refined) existential variable with the
+      (typechecked) original value *)
+      Evarconv.unify_delay !!env sigma newarg (j_val j)
+    in
+    (* We now refine any arguments whose typing was delayed for
+       bidirectionality *)
+    let sigma, t = app in
+    let sigma = List.fold_left refine_arg sigma bidiargs in
```

Comment:
Aren't you refining these `bidiargs` in the wrong order, w.r.t. the order in which there occur in the original term?

Reply:
No problem, now I see the `List.rev`